castra/core.py

```diff
@@ -190,15 +192,31 @@ def unpack_file(fn):
         return np.array(msgpack.unpackb(blosc.decompress(bytes)))
 
 
-def select_partitions(partition_list, key):
+from pandas.core.indexing import convert_to_index_sliceable
+import numpy as np
+
+
+def coerce_index(dt, o):
+    if np.issubdtype(dt, np.datetime64):
+        return pd.Timestamp(o)
+    return o
+
+
+def select_partitions(partitions, key):
     """ Select partitions from partition list given slice
 
-    >>> pl = [(0, 'a'), (10, 'b'), (20, 'c'), (30, 'd'), (40, 'e')]
-    >>> select_partitions(pl, slice(3, 25))
-    (1, 4)
+    >>> p = pd.Series(['a', 'b', 'c', 'd', 'e'], index=[0, 10, 20, 30, 40])
+    >>> select_partitions(p, slice(3, 25))
+    ['b', 'c', 'd']
     """
     assert key.step is None
     start, stop = key.start, key.stop
-    i = bisect(partition_list, (start, None)) if start is not None else None
-    j = bisect(partition_list, (stop, None)) + 1 if stop is not None else None
-    return i, j
+    names = list(partitions.loc[start:stop])
+
+    last = partitions.searchsorted(names[-1])[0]
+
+    stop2 = coerce_index(partitions.index.dtype, stop)
```

Comment:
This uses loc to get out the filenames of most of the partitions.  It then checks the last value to see if we need to add one more.